File: frontend/app/types/portfolio.ts
// Shared types for the application
export interface AccountInfo {
  account_id: string
  status: "ACTIVE" | "INACTIVE"
  currency: string
  cash: string
  cash_withdrawable: string
  cash_transferable: string
  portfolio_value: string
  pattern_day_trader: boolean
  trade_suspended_by_user: boolean
  shorting_enabled: boolean
  long_market_value: string
  short_market_value: string
  equity: string
  last_equity: string
  multiplier: string
}

export interface Position {
  asset_id: string
  symbol: string
  name: string
  qty: string
  avg_entry_price: string
  market_value: string
  cost_basis: string
  unrealized_pl: string
  unrealized_plpc: string
  change_today: string
  asset_class: string
  exchange: string
  id: string
}

export interface Order {
  id: string
  client_order_id: string
  created_at: string
  updated_at: string
  submitted_at: string
  filled_at: string
  expired_at: string | null
  canceled_at: string | null
  failed_at: string | null
  symbol: string
  qty: string
  legs: any
  side: "buy" | "sell"
  type: "market" | "limit"
  trend: string
  time_in_force: string
  limit_price: string | null
  stop_price: string | null
  status: "new" | "partially_filled" | "filled" | "done"
  extended_hours: boolean
  leveraged_by: string | null
  timestamp: string
  filled_avg_price: string
}

export interface CoveredCallOpportunity {
  symbol: string
  underlying_price: number
  option_symbol: string
  strike_price: number
  expiration_date: string
  days_to_expiry: number
  bid: number
  ask: number
  last_price: number
  volume: number
  open_interest: number
  implied_volatility: number
  delta: number
  gamma: number
  theta: number
  vega: number
  premium_received_per_share: number
  total_premium_received: number
  max_return_if_called_away: number
  annualized_return_rate: number
  probability_itm: number
  cagr_if_held_to_expiry: number
  recommendation: "HOLD_POSITION" | "INITIATE_POSITION" | "MONITOR_CLOSELY" | "CLOSE_POSITION"
  reasoning: string
}

export interface PortfolioData {
  account_info: AccountInfo
  positions: Position[]
  orders: Order[]
  covered_call_opportunities: CoveredCallOpportunity[]
}

File: backend/app/routes/portfolio.py
"""GET /portfolio — account + positions from Alpaca paper API (mock fallback)."""

from __future__ import annotations

from fastapi import APIRouter
from ..alpaca_client import AlpacaClient, is_configured
from ..mock_data import mock_account, mock_positions

router = APIRouter()


def _mode() -> str:
    return "live" if is_configured() else "mock"


@router.get("/portfolio")
async def get_portfolio() -> dict:
    if not is_configured():
        return {
            "mode": _mode(),
            "account_info": mock_account(),
            "positions": mock_positions(),
        }
    client = AlpacaClient()
    try:
        account = client.get_account()
    except RuntimeError as exc:
        return {"mode": "error", "detail": str(exc), "account_info": {}, "positions": [], "orders": []}
    try:
        positions = client.get_positions()
    except RuntimeError as exc:
        return {"mode": "error", "detail": str(exc), "account_info": account, "positions": [], "orders": []}
    try:
        orders = client.list_orders(status="all")
    except RuntimeError:
        orders = []
    return {
        "mode": _mode(),
        "account_info": {
            "account_id": account.get("id"),
            "status": account.get("status"),
            "currency": account.get("currency"),
            "cash": account.get("cash"),
            "portfolio_value": account.get("portfolio_value"),
            "equity": account.get("equity"),
            "last_equity": account.get("last_equity"),
            "long_market_value": account.get("long_market_value"),
            "short_market_value": account.get("short_market_value"),
            "pattern_day_trader": account.get("pattern_day_trader"),
            "trade_suspended_by_user": account.get("trade_suspended_by_user"),
            "multiplier": account.get("multiplier"),
        },
        "positions": [
            {
                "symbol": p.get("symbol"),
                "qty": p.get("qty"),
                "avg_entry_price": p.get("avg_entry_price"),
                "market_value": p.get("market_value"),
                "cost_basis": p.get("cost_basis"),
                "unrealized_pl": p.get("unrealized_pl"),
                "unrealized_plpc": p.get("unrealized_plpc"),
                "change_today": p.get("change_today"),
                "asset_class": p.get("asset_class"),
                "exchange": p.get("exchange"),
                "asset_id": p.get("asset_id"),
                "id": p.get("id"),
            }
            for p in positions
        ],
        "orders": orders,
    }
